Google.Api.Gax.Grpc/Rest/RestMethod.cs

```diff
@@ -9,198 +9,65 @@
-        /// Ignores the possibility that the path can already have parameters or contain an anchor (`#`)
-        /// </summary>
-        /// <param name="uriPath">The path component of the service URI</param>
-        /// <param name="queryStringParameters">The parameters to encode in the query string</param>
-        /// <returns>An uri path merged with the encoded query string parameters</returns>
-        private static string AppendQueryStringParameters(string uriPath, IOrderedEnumerable<KeyValuePair<string, string>> queryStringParameters)
+    /// <summary>
+    /// Creates a <see cref="RestMethod"/> representation from the given protobuf method representation.
+    /// </summary>
+    /// <param name="method">The protobuf method to represent.</param>
+    /// <param name="parser">The JSON parser to use when parsing requests.</param>
+    /// <returns>A representation of the method that can be used to handle HTTP requests/responses.</returns>
+    internal static RestMethod Create(MethodDescriptor method, JsonParser parser)
+    {
+        var rule = method.GetOptions().GetExtension(AnnotationsExtensions.Http);
+        if (rule is null)
         {
-            var sb = new StringBuilder();
-            sb.Append(uriPath);
-            bool sbHasParameters = false;
-
-            foreach (var kvpNameValue in queryStringParameters)
-            {
-                sb.Append(sbHasParameters ? "&" : "?");
-                sbHasParameters = true;
-                sb.Append(Uri.EscapeDataString(kvpNameValue.Key));
-                sb.Append("=");
-                sb.Append(Uri.EscapeDataString(kvpNameValue.Value));
-            }
-            
-            return sb.ToString();
+            throw new ArgumentException($"Method {method.Name} in service {method.Service.Name} has no HTTP rule");
         }
+        // TODO: create multiple rules with annotation bindings
+        var transcoder = new HttpRuleTranscoder(method, rule);
+        return new RestMethod(method, parser, transcoder);
+    }
 
-        // TODO: Handle cancellation?
-        /// <summary>
-        /// Parses the response and converts it into the protobuf response type.
-        /// </summary>
-        internal async Task<TResponse> ReadResponseAsync<TResponse>(Task<ReadHttpResponseMessage> httpResponseTask)
-        {
-            var httpResponse = await httpResponseTask.ConfigureAwait(false);
-            var status = httpResponse.GetStatus();
-            if (status.StatusCode != StatusCode.OK)
-            {
-                throw new RpcException(status, httpResponse.GetTrailers());
-            }
-            return (TResponse) _parser.Parse(httpResponse.Content, _protoMethod.OutputType);
-        }
+    internal HttpRequestMessage CreateRequest(IMessage request, string host)
+    {
+        // TODO: Use multiple rules, iterating over them until a match is found.
+        var output = _transcoder.Transcode(request);
+        return output.CreateRequest(host);
```

Comment:
nit: Could these be squashed into one method call `_transcoder.CreateTranscodedRequest(request, how)`? Even if `TranscodingOutput` remains a class (maybe a child class of `HttpRuleTranscoder`).

Or is the separetion useful for testing or something else?

Reply:
It's useful for #579 :)